arkanoid/js/game.js

```diff
@@ -0,0 +1,219 @@
+    var loaded = 0;
+    var totalSpritesLength = Object.getOwnPropertyNames(this.sprites).length;
+    var totalSoundsLength = Object.getOwnPropertyNames(this.sounds).length;
+    var totalForPreload = totalSpritesLength + totalSoundsLength;
+
+    var onResourceLoad = function () {
+      ++loaded;
+      if (loaded >= totalForPreload) {
+        callback();
+      }
+    };
+
+    this.loadSprites(onResourceLoad);
+    this.loadSounds(onResourceLoad);
+  };
+
+  this.loadSprites = function (callback) {
+    for (var key in this.sprites) {
+      this.sprites[key] = new Image();
+      this.sprites[key].src = "assets/img/" + key + ".png";
+      this.sprites[key].addEventListener("load", callback);
+    }
+  };
+
+  this.loadSounds = function (callback) {
+    for (var key in this.sounds) {
+      this.sounds[key] = new Audio("assets/sounds/" + key + ".mp3");
+      this.sounds[key].addEventListener("canplaythrough", callback, { once: true });
+    }
+  };
+
+  this.update = function () {
+    this.collideBlocks();
+    this.collidePlatform();
+    this.ball.collideWorldBounds(function () {
+      me.end("gameOver");
+    });
+    this.platform.collideWorldBounds(this.width);
+    this.platform.move();
+    this.ball.fly();
+  };
+
+  this.clear = function () {
+    this.ctx.clearRect(0, 0, this.width, this.height);
+  };
+
+  this.render = function () {
+    this.ctx.save();
+    this.ctx.drawImage(this.sprites.ball, this.ball.x, this.ball.y);
+    this.ctx.drawImage(this.sprites.platform, this.platform.x, this.platform.y);
+    this.renderText("Score: " + this.score);
+    this.renderBlocks();
+    this.ctx.restore();
+  };
+
+  this.renderText = function (text) {
+    this.ctx.fillText(text, 10, 20);
+  };
+
+  this.createBlocks = function () {
+    for (var row = 0; row < this.rows; row++) {
+      for (var col = 0; col < this.cols; col++) {
+        this.blocks.push(new Block(col, row));
+      }
+    }
+  };
+
+  this.renderBlocks = function () {
+    for (var block in this.blocks) {
+      var currentBlock = this.blocks[block];
+      if (currentBlock.active) {
+        this.ctx.drawImage(this.sprites.block, currentBlock.x, currentBlock.y);
+      }
+    }
+  };
+
+  this.collideBlocks = function () {
+    for (var block in this.blocks) {
+      var currentBlock = this.blocks[block];
+      if (currentBlock.active && this.ball.collide(currentBlock)) {
+        this.ball.bumpBlock(currentBlock);
+        this.playSound("bump");
+        this.addScore();
+      }
+    }
+  };
+
+  this.collidePlatform = function () {
+    if (this.ball.collide(this.platform)) {
+      this.ball.bumpPlatform(this.platform);
+      this.playSound("jump");
+    }
+  };
+
+  this.addScore = function () {
+    ++this.score;
+    if (this.score >= this.blocks.length) {
+      this.end("winner");
+    }
+  };
+
+  this.run = function () {
+    if (this.running) {
+      window.requestAnimationFrame(function () {
+        me.update();
```

Comment:
Methods call/declaration order. Consider having the method declared directly after it's being called.
In this specific case, "update" is declared way before before it's called:
called on line 181, declared on line 108. This makes code less readable (the one who sees it for the first time is not able to understand what it's used for without extra effort of searching it in the code).